### ISSUE TITLE
M 2.2.6 - Incomplete/Partial Product Price Reindex on catalog rule save.

### ISSUE THREAD

**Issue by neodrg** [magento/magento2#18305]
### Preconditions

1.Magento 2.2.6

### Steps to reproduce

1.Create a configurable product with 3 simple product link (3 colors)
2. add price 45 on each simple product
3. add general - retailer price on advanced price retailer (45.0000)
4. create a catalog rule on retailer customer group
4.1 conditions SKU is one of (select the three simple product)
4.2 Actions : adjust final price to discount value - discount amount 38.0000
4.3 discard subsequent rules : NO
5 reindex 

### Expected result

1.on front all item have a price of 38 and old price of 45 with line-through

### Actual result

1.on front all item has a price of 38 and old price of 45 with line-through
2.first item is good
3. second item is always with 45 no old price.
![capture1](https://user-images.githubusercontent.com/7758910/46213151-1e601b80-c338-11e8-8ab3-542284b876ff.PNG)
![capture3](https://user-images.githubusercontent.com/7758910/46213152-1e601b80-c338-11e8-8ef9-27b7842cbfd2.PNG)

It seems the data is not re-index correctly because in DB
First item final price is good - second and third is not. 
![capture4](https://user-images.githubusercontent.com/7758910/46213296-7d259500-c338-11e8-9110-3556110f01b4.PNG)

If you go then on simple product page and then save again price has been changed in DB correctly.

![capture5](https://user-images.githubusercontent.com/7758910/46213560-279db800-c339-11e8-9b83-0e3fd7f6e987.PNG)

If you change catalog rule adjust final price to 39 then only the first item is changing.

A complete re-index not resolving the issue.

This was working on 2.2.3

**Comment by magento-engcom-team** [comment]
Hi @neodrg. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [ ] Summary of the issue
- [ ] Information on your environment
- [ ] Steps to reproduce
- [ ] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@neodrg do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [ ] yes
- [ ] no

**Comment by neodrg** [comment]
@magento-engcom-team give me 2.2.6 instance

**Comment by magento-engcom-team** [comment]
Hi @neodrg. Thank you for your request. I'm working on Magento 2.2.6 instance for you

**Comment by magento-engcom-team** [comment]
Hi @neodrg, here is your [Magento instance](https://i-18305-2-2-6.engcom.dev.magento.com).
Admin access: [https://i-18305-2-2-6.engcom.dev.magento.com/admin](https://i-18305-2-2-6.engcom.dev.magento.com/admin)
Login: `admin` Password: `123123q`
Instance will be terminated in up to 3 hours.

**Comment by neodrg** [comment]
@magento-engcom-team Please re-index my instance

**Comment by neodrg** [comment]
Personal Investigation :
I made some test on catalog rule.
There is an error with the operator "is one of", only the first item specified is update in the table  catalog_product index_price in DB.

**Comment by hatimeria-artur-jewula** [comment]
I stumble on similar problem with price indexer. It looks like price indexer no longer reindex product types according to indexPriority. Up to version 2.2.5 price indexer run with list od ids to reindex sorted them to composite and notComposite and later run product type indexers in order set by `indexPriority` (https://github.com/magento/magento2/blob/2.2.5/app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php#L318) but in version 2.2.6 (https://github.com/magento/magento2/blob/2.2.6/app/code/Magento/Catalog/Model/Indexer/Product/Price/AbstractAction.php#L364) this has changes and now product type price indexers are run in order based on the order of product ids passed to `_reindexRows` function. This function first removes old values in index table for all ids and later reindex products by type. If a type indexer relies on another type to be finished earlier (ie. configurable min and max price uses catalog_product_index_price) table. Then it may happen that it will not get proper results. The fix is to make sure `for` loop is executed in order of `indexPriority`

**Comment by neodrg** [comment]
@hatimeria-artur-jewula "The fix is to make sure for loop is executed in order of indexPriority"
How would you make this change?

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-andrii. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `Issue: Confirmed` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-andrii Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-96609`, `MAGETWO-96610` were created

**Comment by jmtakahashi** [comment]
A complete failure for all America based stores trying to use this for Black Friday.  My store also has this issue.

**Comment by magento-engcom-team** [comment]
@neodrg, thank you for your report.     
    
Unfortunately, we are archiving this ticket now as it did not get much attention from both Magento Community and Core developers for an extended period. This is done in an effort to create a quality, community-driven backlog which will allow us to allocate the required attention more easily.

Please feel free to comment, reopen or create new ticket according to the [Issue reporting guidelines](https://github.com/magento/magento2/wiki/Issue-reporting-guidelines) 
if you are still facing this issue on the latest `2.3-develop` branch. Thank you for collaboration.
